packages/phishing-controller/src/PhishingDetector.ts

```diff
@@ -0,0 +1,177 @@
+import { distance } from 'fastest-levenshtein';
+
+import {
+  domainPartsToDomain,
+  domainPartsToFuzzyForm,
+  domainToParts,
+  getDefaultPhishingDetectorConfig,
+  matchPartsAgainstList,
+  processConfigs,
+} from './utils';
+
+export type LegacyPhishingDetectorList = {
+  whitelist?: string[];
+  blacklist?: string[];
+} & FuzzyTolerance;
+
+export type PhishingDetectorList = {
+  allowlist?: string[];
+  blocklist?: string[];
+  name?: string;
+  version?: string | number;
+} & FuzzyTolerance;
+
+export type FuzzyTolerance =
+  | {
+      tolerance?: number;
+      fuzzylist: string[];
+    }
+  | {
+      tolerance?: never;
+      fuzzylist?: never;
+    };
```

Comment:
What was your intent with using `never`, to indicate that these shouldn't be provided or something else? Curiously, if I hover over this type, then TypeScript converts the `never`s to `undefined`:

``` typescript
export type FuzzyTolerance =
  | {
      tolerance?: number;
      fuzzylist: string[];
    }
  | {
      tolerance?: never;
      fuzzylist?: never;
    };
```

However, I'm curious whether the `never`s are needed at all. The `?` already means that these properties are optional, so would this work?

``` suggestion
export type FuzzyTolerance = {
  tolerance?: number;
  fuzzylist: string[];
};
```

Reply:
Good point. My intent here was to communicate through types that tolerance is used only in the presence of a fuzzylist, and in case of a missing fuzzylist (but with tolerance) the detector throws an error.

So, either the user passes a fuzzylist and _maybe_ a tolerance, or neither of the two.

With this change it would work:
```suggestion
export type FuzzyTolerance =
  | {
      tolerance?: number;
      fuzzylist: string[];
    }
  | {
      tolerance: never;
      fuzzylist: never;
    };
 ```
 
 Let me know if this makes sense

Reply:
Gotcha. Okay, that makes sense.